Add Healthcare beta API and associated resources and iam

yinzara opened the pull request:
The [Google Healthcare API](https://cloud.google.com/healthcare/) is now in beta release. This PR adds all support for datasets, fhirStores, dicomStores, and hl7v2Stores including iam policies, members and bindings for all 4.

I was careful to wrap all classes as erb(s) with the appropriate exclusions to ensure it only appears in the terraform-provider-google-beta.

There are test to cover all healthcare resources and iam resources.

You will need to add the dependency of:
google.golang.org/api/healthcare/v1beta1
(I was unsure of how to specify that in Magic Modules)

# Release Note for Downstream PRs
```releasenote
* **New Resource** `google_healthcare_dataset_iam_binding`
* **New Resource** `google_healthcare_dataset_iam_member`
* **New Resource** `google_healthcare_dataset_iam_policy`
* **New Resource** `google_healthcare_dicom_store_iam_binding`
* **New Resource** `google_healthcare_dicom_store_iam_member`
* **New Resource** `google_healthcare_dicom_store_iam_policy`
* **New Resource** `google_healthcare_fhir_store_iam_binding`
* **New Resource** `google_healthcare_fhir_store_iam_member`
* **New Resource** `google_healthcare_fhir_store_iam_policy`
* **New Resource** `google_healthcare_hl7_v2_store_iam_binding`
* **New Resource** `google_healthcare_hl7_v2_store_iam_member`
* **New Resource** `google_healthcare_hl7_v2_store_iam_policy`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

yinzara commented:
Didn't quite understand how the testing works but now that I do I was able to fix all the tests and cover all functionality added.

1. Generate terraform-provider-google-beta 
`bundle exec compiler -a -v "beta" -e terraform -o "$GOPATH/src/github.com/terraform-providers/terraform-provider-google-beta"`

2.`cd $GOPATH/src/github.com/terraform-providers/terraform-provider-google-beta`
3. `make && make test`
4. `make testacc TEST="./google-beta" TESTARGS='-run=TestAccHealthcareHl7V2Store'`

danawillow commented:
drive-by: @yinzara, thanks so much for this hefty contribution! In the future, if you don't mind, having small PRs that each accomplish a self-contained purpose would go a long way in helping us review your code faster (and of course I hope you continue contributing 😄 )

yinzara commented:
Understood :) I really only needed control of the FHIR store resource but felt if I just contributed the code to make that work, it would have been an incomplete submission so I implemented the entirety of the healthcare API.

chrisst commented:
@yinzara First of all, this is a ton of work and we really want to thank you for putting this work in! To help offset your cost of integration tests we'd be happy to extend some GCP credits. Please email me chrisst@google for more details.

Secondly as Dana mentioned the size of this PR means it has taken me a while to find a block of time to review it in detail. I'll be working on it this week and will be quicker to respond after I've done an initial pass.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of 38b909162a027866d3b2db7b7613cda193daea3e. 
## Pull request statuses
No diff detected in terraform-google-conversion.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/858 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/3870

chrisst commented on products/healthcare/terraform.yaml:
```diff
@@ -0,0 +1,103 @@
+    import_format: ["{{dataset}}/dicomStores/{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "healthcare_dicom_store_basic"
```
```suggestion
        name: "healthcare_dicom_store_basic"
        skip_test: true
```

yinzara replied:
Why should I be skipping tests for this? Is it because you're suggesting I wrote update tests that would probably cover this?

chrisst replied:
Ah - I should have explained better. If you exclude import the tests will fail to generate. So the two must be done in tandem. Until I fix the test generation so it will create valid tests :P

yinzara replied:
Well I don't have ass anything so I wasn't about to enable "exclude_import" :-) . I've written the custom importers for the 3 stores.  I'm just working on the integration tests now (admittedly using the generated ones as the base to test changes was helpful).

chrisst commented on products/healthcare/terraform.yaml:
```diff
@@ -0,0 +1,103 @@
+    description: |
+      {{description}}
+    id_format: "{{dataset}}/dicomStores/{{name}}"
+    import_format: ["{{dataset}}/dicomStores/{{name}}"]
```
The import regex generator currently only matches fields that are in between slashes and doesn't do well with variables that contain slashes.
see https://github.com/terraform-providers/terraform-provider-google-beta/pull/858/files#diff-14199601f6d40245f2b47f254e2722deR217

We would usually set it to something like `"projects/{{project}}/locations/{{location}}/datasets/{{dataset}}/dicomStores/{{name}}"` but in this case there would need to be settable values for `project` and `location` which for these resources is not needed.

This is a long way of saying that magic modules doesn't currently have adequate import support for this type of resource without adding a `custom_import` method here. You are welcome to add custom import support if you need it or you can exclude import for now and wait until we fix https://github.com/GoogleCloudPlatform/magic-modules/issues/1946

```suggestion
    exclude_import: true
```

yinzara commented:
Should I fix the original PR from Magic Modules or fix the generated code here?

chrisst commented:
The original MM PR is the right place to fix. Which I think is this PR but if not can you link me to what you are talking about?

yinzara commented:
Apologies. I thought I was in the other generated PR inside the google-cloud-provider-beta. Ignore me :)

yinzara commented on third_party/terraform/utils/iam_healthcare_dataset.go.erb:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (u *HealthcareDatasetIamUpdater) DescribeResource() string {
+	return fmt.Sprintf("KMS Dataset %q", u.resourceId)
```
You caught my copy and paste error! Dang! I was so close!

yinzara commented:
I'm nearly done with my fixes. Do you want me to push them as a new commit first then rebase against master once you've approved or should I rebase against master now? Or do we not rebase against master in this project?

chrisst commented:
Feel free to rebase against master whenever you want. Our build system will rebase the changes during the process of merging the PR's so it's perfectly fine to do it whenever.

yinzara commented:
So it seems I've found a bug in the actual real Google Healthcare Fhir Store API :) For one of the parameters of a FHIR store (enableHistoryImport), though the PATCH request accepts an update and the response includes the updated value, the response from the corresponding GET does not include parameter making my tests fail. 

See the following log:
```
---[ REQUEST ]---------------------------------------
PATCH /v1beta1/projects/mycareplan-dev/locations/us-central1/datasets/tf-test-e98bnv90hk/fhirStores/tf-test-yddifz1ja4?alt=json&updateMask=enableUpdateCreate%2CenableHistoryImport%2Clabels%2CnotificationConfig HTTP/1.1
Host: healthcare.googleapis.com
User-Agent: Terraform/0.12.2 (+https://www.terraform.io) terraform-provider-google-beta/acc
Content-Length: 178
Content-Type: application/json
Accept-Encoding: gzip

{
 "enableHistoryImport": true,
 "enableUpdateCreate": true,
 "labels": {
  "label1": "labelvalue1"
 },
 "notificationConfig": {
  "pubsubTopic": "projects/mycareplan-dev/topics/tf-test-k70vown181"
 }
}

-----------------------------------------------------
2019/06/21 11:49:40 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Fri, 21 Jun 2019 18:49:12 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/mycareplan-dev/locations/us-central1/datasets/tf-test-e98bnv90hk/fhirStores/tf-test-yddifz1ja4",
  "enableUpdateCreate": true,
  "notificationConfig": {
    "pubsubTopic": "projects/mycareplan-dev/topics/tf-test-k70vown181"
  },
  "labels": {
    "label1": "labelvalue1"
  },
  "enableHistoryImport": true
}

-----------------------------------------------------
2019/06/21 11:49:40 [DEBUG] Waiting for state to become: [success]
2019/06/21 11:49:40 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1beta1/projects/mycareplan-dev/locations/us-central1/datasets/tf-test-e98bnv90hk/fhirStores/tf-test-yddifz1ja4?alt=json HTTP/1.1
Host: healthcare.googleapis.com
User-Agent: Terraform/0.12.2 (+https://www.terraform.io) terraform-provider-google-beta/acc
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------
2019/06/21 11:49:40 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: quic=":443"; ma=2592000; v="46,44,43,39"
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Fri, 21 Jun 2019 18:49:13 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/mycareplan-dev/locations/us-central1/datasets/tf-test-e98bnv90hk/fhirStores/tf-test-yddifz1ja4",
  "enableUpdateCreate": true,
  "notificationConfig": {
    "pubsubTopic": "projects/mycareplan-dev/topics/tf-test-k70vown181"
  },
  "labels": {
    "label1": "labelvalue1"
  }
}
```

yinzara commented:
I've pushed an update that implements all your changes from the review.  I also have a separate branch prepared that has been rebased against master with the original submitted commit and changes squashed into a single commit. If you'd prefer I push that one, I will do that or you can approve and merge this one (or make more suggestions :))

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, d54f50a812e109217b6c076b0e6367e48a7b1272. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in terraform-google-conversion.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

chrisst commented:
The tests are failing right now with:
```

chrisst commented:
Ah thanks, I was halfway through writing a follow up comment saw you rebased already! I'm running through our build system again now.

yinzara commented:
lol keeping up with changes to magic modules is a full time job ;-) . Guess that function was refactored from when I branched.  I also found the 'default' parameter on the API versions in api.yaml is now not supported (so I removed it).  I've merged to master and fixed issues.

yinzara commented:
I was noticing something else very odd.  All "IamMember" related tests across the entire repository are now failing for me. 
For example:
`make testacc TEST=./google-beta TESTARGS='-run=TestAccComputeSubnetworkIamMember'`
now fails with the error:
```==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google-beta -v -run=TestAccComputeSubnetworkIamMember -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google-beta/version.ProviderVersion=acc"
=== RUN   TestAccComputeSubnetworkIamMember
=== PAUSE TestAccComputeSubnetworkIamMember
=== CONT  TestAccComputeSubnetworkIamMember
--- FAIL: TestAccComputeSubnetworkIamMember (88.59s)
    testing.go:568: Step 1 error: Failed state verification, resource with ID projects/mycareplan-dev/regions/us-central1/subnetworks/tf-test-p98x4m70xx/roles/compute.networkUser/serviceAccount:tf-test-7925644170588690737@mycareplan-dev.iam.gserviceaccount.com not found
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google-beta/google-beta	88.627s
make: *** [testacc] Error 1
```

They all passed when I last ran them but somehow they're now failing. All the rest of the IAM tests are fine (i.e. policy and binding) but member fails consistently across every product.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 4b0b676663a7b914658886259bdf433cec2f4955. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in terraform-google-conversion.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

chrisst commented:
Lol, the churn isn't always this high but it's been a busy few weeks.

The IAM integration tests are failing due to a different change that just got merged in yesterday and the fix is coming in https://github.com/GoogleCloudPlatform/magic-modules/pull/1975

chrisst commented:
Our build system can't merge this right now because there are a few rebase conflicts that will need to be fixed by hand. I don't have permission to update your branch so I've pushed up a clean up branch at https://github.com/chrisst/magic-modules/tree/yinza-healthcare . Can you either give me permission on your fork temporarily or force push from my fork's branch to this master?

yinzara commented:
You're all set! Thanks!

yinzara commented:
Well build is still having issues. I've given you access to my fork so you can push to your hearts content!

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of d9c4a124af2a51af3892576f58a146c1afa1afdf. 
## Pull request statuses
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/898

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 35843290dae867562b68cb1cf55ddb1d3b511f1b. 
## Pull request statuses
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/899

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 35843290dae867562b68cb1cf55ddb1d3b511f1b. 
## Pull request statuses
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/900

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b9dfba661dde15430088e1d2cbf099f09673cef0. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b9dfba661dde15430088e1d2cbf099f09673cef0. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in terraform-google-conversion.
No diff detected in terraform-provider-google.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.
